x-pack/plugins/visualization_lens/public/types.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+export interface DatasourceSuggestion<T> {
+  state: T;
+  tableMetas: TableMetaInfo[];
+}
+
+/**
+ * Interface for the datasource registry
+ */
+export interface Datasource<T> {
+  // For initializing from saved object
+  // initialize: (state?: T) => Promise<T>;
+
+  renderDataPanel: (props: DatasourceDataPanelProps) => void;
+
+  toExpression: (state: T) => string;
+
+  getDatasourceSuggestionsForField: (state: T) => Array<DatasourceSuggestion<T>>;
+  getDatasourceSuggestionsFromCurrentState: (state: T) => Array<DatasourceSuggestion<T>>;
+
+  getPublicAPI: (state: T, setState: (newState: T) => void) => DatasourcePublicAPI;
+  // publicAPI: {
+  //   [key in keyof DatasourcePublicAPI]: (
+  //     state: T,
+  //     setState: (newState: T) => void
+  //   ) => DatasourcePublicAPI[key]
+  // };
+}
+
+/**
+ * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
+ */
+export interface DatasourcePublicAPI {
+  getTableSpec: () => TableSpec;
+  getOperationForColumnId: () => Operation;
+
+  // Called by dimension
+  getDimensionPanelComponent: (
+    props: DatasourceDimensionPanelProps
+  ) => (domElement: Element, operations: Operation[]) => void;
+
+  removeColumnInTableSpec: (columnId: string) => TableSpec;
+  moveColumnTo: (columnId: string, targetIndex: number) => void;
+  duplicateColumn: (columnId: string) => TableSpec;
+}
+
+export interface DatasourceDataPanelProps {
+  domElement: Element;
+}
+
+// The only way a visualization has to restrict the query building
+export interface DatasourceDimensionPanelProps {
+  // If no columnId is passed, it will render as empty
+  columnId?: string;
+
+  // Visualizations can restrict operations based on their own rules
+  filterOperations: (operation: Operation) => boolean;
+
+  // Visualizations can hint at the role this dimension would play, which
+  // affects the ordering of the query
+  suggestedPriority?: DimensionPriority;
+}
+
+export interface DatasourceValidationError {
+  type: 'error';
+}
+
+export interface DatasourceValidationValid {
+  type: 'valid';
+}
+
+export type DataType = 'string' | 'number' | 'date' | 'boolean';
+
+// An operation does not represent a column in the datatable
+export interface Operation {
+  // Operation ID is a reference to the operation
+  id: string;
+  // User-facing label for the operation
+  label: string;
+  dataType: DataType;
+  // A bucketed operation has many values the same
+  isBucketed: boolean;
+
+  // Extra meta-information like cardinality, color
+}
+
+export interface TableSpecColumn {
+  // Column IDs are the keys for internal state in data sources and visualizations
+  columnId: string;
+}
+
+// TableSpec is managed by visualizations
+export type TableSpec = TableSpecColumn[];
+
+export type VisualizationTableRequest = Array<{
+  dataType: 'string' | 'number';
+  isBucketed: boolean;
+}>;
+
+export interface VisualizationProps<T> {
+  datasource: DatasourcePublicAPI;
```

Comment:
I forget where we ended up in our debate about whether or not this should be plural. Just want to double-check that we decided it should be singular.